File: Habitica/src/main/java/com/habitrpg/android/habitica/data/SocialRepository.java
package com.habitrpg.android.habitica.data;

import com.habitrpg.android.habitica.models.AchievementResult;
import com.habitrpg.android.habitica.models.inventory.Quest;
import com.habitrpg.android.habitica.models.members.Member;
import com.habitrpg.android.habitica.models.responses.PostChatMessageResult;
import com.habitrpg.android.habitica.models.social.Challenge;
import com.habitrpg.android.habitica.models.social.ChatMessage;
import com.habitrpg.android.habitica.models.social.Group;
import com.habitrpg.android.habitica.models.user.User;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.realm.RealmResults;
import rx.Observable;

public interface SocialRepository extends BaseRepository {
    Observable<List<ChatMessage>> retrieveGroupChat(String groupId);
    Observable<RealmResults<ChatMessage>> getGroupChat(String groupId);

    void markMessagesSeen(String seenGroupId);

    Observable<Void> flagMessage(ChatMessage chatMessage);

    Observable<ChatMessage> likeMessage(ChatMessage chatMessage);

    Observable<Void> deleteMessage(ChatMessage chatMessage);

    Observable<PostChatMessageResult> postGroupChat(String groupId, HashMap<String, String> messageObject);
    Observable<PostChatMessageResult> postGroupChat(String groupId, String message);

    Observable<Group> retrieveGroup(String id);
    Observable<Group> getGroup(String id);

    Observable<Group> leaveGroup(String id);

    Observable<Group> joinGroup(String id);

    Observable<Void> updateGroup(Group group, String name, String description, String leader, String privacy);

    Observable<List<Group>> retrieveGroups(String type);
    Observable<RealmResults<Group>> getGroups(String type);
    Observable<RealmResults<Group>> getPublicGuilds();

    Observable<PostChatMessageResult> postPrivateMessage(HashMap<String, String> messageObject);
    Observable<PostChatMessageResult> postPrivateMessage(String recipientId, String message);


    Observable<RealmResults<Member>> getGroupMembers(String id);
    Observable<List<Member>> retrieveGroupMembers(String id, boolean includeAllPublicFields);

    Observable<Void> inviteToGroup(String id, Map<String, Object> inviteData);

    Observable<List<Challenge>> getUserChallenges();

    Observable<Member> getMember(String userId);

    Observable<Void> markPrivateMessagesRead(User user);

    Observable<RealmResults<Group>> getUserGroups();

    Observable<Void> acceptQuest(User user, String partyId);
    Observable<Void> rejectQuest(User user, String partyId);

    Observable<Void> leaveQuest(String partyId);

    Observable<Void> cancelQuest(String partyId);

    Observable<Quest> abortQuest(String partyId);

    Observable<Void> rejectGroupInvite(String groupId);

    Observable<Quest> forceStartQuest(Group party);

    Observable<AchievementResult> getMemberAchievements(String userId);
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/data/SocialRepository.kt
package com.habitrpg.android.habitica.data

import com.habitrpg.android.habitica.models.Achievement
import com.habitrpg.android.habitica.models.inventory.Quest
import com.habitrpg.android.habitica.models.invitations.InviteResponse
import com.habitrpg.android.habitica.models.members.Member
import com.habitrpg.android.habitica.models.responses.PostChatMessageResult
import com.habitrpg.android.habitica.models.social.ChatMessage
import com.habitrpg.android.habitica.models.social.FindUsernameResult
import com.habitrpg.android.habitica.models.social.Group
import com.habitrpg.android.habitica.models.social.GroupMembership
import com.habitrpg.android.habitica.models.social.InboxConversation
import com.habitrpg.android.habitica.models.user.User
import io.realm.RealmResults
import kotlinx.coroutines.flow.Flow

interface SocialRepository : BaseRepository {
    fun getUserGroups(type: String?): Flow<List<Group>>

    suspend fun retrieveGroupChat(groupId: String): List<ChatMessage>?

    fun getGroupChat(groupId: String): Flow<List<ChatMessage>>

    suspend fun markMessagesSeen(seenGroupId: String)

    suspend fun flagMessage(
        chatMessageID: String,
        additionalInfo: String,
        groupID: String? = null,
    ): Void?

    suspend fun reportMember(
        memberID: String,
        data: Map<String, String>,
    ): Void?

    suspend fun likeMessage(chatMessage: ChatMessage): ChatMessage?

    suspend fun deleteMessage(chatMessage: ChatMessage): Void?

    suspend fun postGroupChat(
        groupId: String,
        messageObject: HashMap<String, String>,
    ): PostChatMessageResult?

    suspend fun postGroupChat(
        groupId: String,
        message: String,
    ): PostChatMessageResult?

    suspend fun retrieveGroup(id: String): Group?

    fun getGroup(id: String?): Flow<Group?>

    suspend fun leaveGroup(
        id: String?,
        keepChallenges: Boolean,
    ): Group?

    suspend fun joinGroup(id: String?): Group?

    suspend fun createGroup(
        name: String?,
        description: String?,
        leader: String?,
        type: String?,
        privacy: String?,
        leaderCreateChallenge: Boolean?,
    ): Group?

    suspend fun updateGroup(
        group: Group?,
        name: String?,
        description: String?,
        leader: String?,
        leaderCreateChallenge: Boolean?,
    ): Group?

    fun getInboxMessages(replyToUserID: String?): Flow<RealmResults<ChatMessage>>

    suspend fun retrieveInboxMessages(
        uuid: String,
        page: Int,
    ): List<ChatMessage>?

    suspend fun retrieveInboxConversations(): List<InboxConversation>?

    fun getInboxConversations(): Flow<RealmResults<InboxConversation>>

    suspend fun postPrivateMessage(
        recipientId: String,
        messageObject: HashMap<String, String>,
    ): List<ChatMessage>?

    suspend fun postPrivateMessage(
        recipientId: String,
        message: String,
    ): List<ChatMessage>?

    suspend fun getPartyMembers(id: String): Flow<List<Member>>

    suspend fun getGroupMembers(id: String): Flow<List<Member>>

    suspend fun retrievePartyMembers(
        id: String,
        includeAllPublicFields: Boolean,
    ): List<Member>?

    suspend fun inviteToGroup(
        id: String,
        inviteData: Map<String, Any>,
    ): List<InviteResponse>?

    suspend fun retrieveMember(
        userId: String?,
        fromHall: Boolean = false,
    ): Member?

    suspend fun findUsernames(
        username: String,
        context: String? = null,
        id: String? = null,
    ): List<FindUsernameResult>?

    suspend fun markPrivateMessagesRead(user: User?)

    fun markSomePrivateMessagesAsRead(
        user: User?,
        messages: List<ChatMessage>,
    )

    suspend fun transferGroupOwnership(
        groupID: String,
        userID: String,
    ): Group?

    suspend fun removeMemberFromGroup(
        groupID: String,
        userID: String,
    ): List<Member>?

    suspend fun acceptQuest(
        user: User?,
        partyId: String = "party",
    ): Void?

    suspend fun rejectQuest(
        user: User?,
        partyId: String = "party",
    ): Void?

    suspend fun leaveQuest(partyId: String): Void?

    suspend fun cancelQuest(partyId: String): Void?

    suspend fun abortQuest(partyId: String): Quest?

    suspend fun rejectGroupInvite(groupId: String): Void?

    suspend fun forceStartQuest(party: Group): Quest?

    suspend fun getMemberAchievements(userId: String?): List<Achievement>?

    suspend fun transferGems(
        giftedID: String,
        amount: Int,
    ): Void?

    fun getGroupMembership(id: String): Flow<GroupMembership?>

    fun getGroupMemberships(): Flow<List<GroupMembership>>

    suspend fun blockMember(userID: String): List<String>?

    fun getMember(userID: String?): Flow<Member?>

    suspend fun updateMember(
        memberID: String,
        data: Map<String, Map<String, Boolean>>,
    ): Member?

    suspend fun retrievePartySeekingUsers(page: Int = 0): List<Member>?

    suspend fun retrievegroupInvites(
        id: String,
        includeAllPublicFields: Boolean,
    ): List<Member>?
}
